EIPS/eip-3712.md

```diff
@@ -0,0 +1,321 @@
+---
+eip: 
+title: ERC Standard for Multiple Types of Fungible-Tokens
+author: Zhao Li(@1999321)、Yuefei Tan (@whtyfhas)、Derek Zhou (@zhous)  
+discussions-to:
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-08-01
+---
+
+## Simple Summary
+
+A standard that manages multiple types of fungible tokens in a single contract.
+
+## Abstract
+
+This standard covers some situations that the ERC-20 standard and ERC-1155 standard can hardly handle. It allows multiple types of fungible tokens to be sent to multiple receiving addresses via our proposed interface `transferBatchMul` and allows multip le types of fungible tokens to be sent from multiple sending addresses to multiple receiving addresses via our proposed interface `transferFromBatchMul`. This standard greatl y reduces the gas consumption in transactions of multiple types of fungible tokens. In addition the standard allows both setting of an approval for a single token via an interface `approve(uint256 id,address spender, uint256 amount)` and setting of approvals uniformly for all types of  tokens via an interface `approve(address spender,bool _status)`.
+
+## Motivation
+
+The ERC-20 token standard defines a single fungible token’s attributes and interfaces. When interacting with multiple types of fungible tokens, multiple token contracts each of which defines a single token need to be developed and deployed. The ERC-1155 token standard combines both fungible tokens and non-fungible tokens but lacks both an interface to set an approval for a spender for an id specified type of fungible token and an interface to set an approval for multiple spenders for an id specified type of fungible token. In addition both the ERC-20 and the ERC-1155 token standards lack interfaces to transact from multiple sending addresses to multiple receiving addresses. This proposed standard covers these use cases and allows multiple types of fungible tokens to be transacted from multiple sending addresses to multiple receiving addresses.   
```

Comment:
It might be worth noting that 1155 technically already supports fungible tokens. This may be an incorrect assumption on my part, but I would assume the reason that 1155 never really took off for ERC20-like tokens is due to the difficulty of integrating them on the smart contract/app layer - web3 code assumes that ERC20-like tokens are one to a contract (and therefore do not need a `tokenId` when being referenced), as do Solidity functions integrating with them. For example, Uniswap (to the best of my knowledge) does not have any ability to integrate a token with a `tokenId` into the pool.

As a result, imho it would be hard to see this EIP getting wide traction in the form you intend (not unlike 1155). I don't mean this to be discouraging - I think the functions that you propose are very useful, and in fact know of contracts implementing similar functions for convenience. imho you might want to change the motivation to reflect that.

Reply:
I'm a team member of this EIP. The application goal (www.DAism.io) of this EIP is to build a platform for Ethereum economy, by creating a band new way to issue tokens and a band new decentralized exchange network (we call it IADD), which is obviously a replacement of Uniswap. 
On Nov 13th, 2019,  I wrote an article relating to NFT (https://deme.app/full#12), But I emilinated it from the core of DAism the next year. So we only focus on fungile token now.
It's a very good suggestion to modify the motivation. Thank you very much!

Reply:
Thank you all for your inputs. Yes what "wschwab" said pretty makes sense. However when utilizing this standard we can handling multiple fungible tokens in a single contract, and it makes transactions of multiple fungible tokens much easier and more gas-efficient. In addition ERC-1155 is more widely used in games while this standard can be more widely used in DeFi. This standard partly inherits some functions in transaction from ERC-1155 but has new functions specifically to handle multipe fungile tokens. ERC-1155 doesn't differentiate fungible tokens from non fungible tokens, doesn't have attributes such as precisions, token name, token symbol to define a fungible token, and doesn't have functions to handle different kinds of approvals for fungible tokens (it only has a function to set a global approval). ERC-1155's overall design is more intended for games. It doesn't cover many use cases such as DeFi in which multiple fungible tokens need to be handled more efficiently. Therefore we come up with our EIP to cover these use cases. We inherit the functions and features of ERC-1155, which apply to fungible tokens and add more attributes such as ID and functions such as different approval functions to better handle multiple fungible tokens in more general use cases. Our proposed EIP is intended for handling multiple fungible tokens in wider scenarios and in a more efficient way.